### PR TITLE
[Eventhubs] Use InMemoryCheckpointStore internally when no checkpoint store from user

### PR DESCRIPTION
**Pull request by YijunXieMS**
1. Deleted sqlite3 based InMemoryCheckpointStore. Added one using Python's built-in dict type. Internal data structure is Trie.
2. Changed EventProcessor (both sync and async) to use it when users don't specify a checkpoint store

### CODE REVIEW

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

File: sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    def __init__(self, value, is_leaf):
+        self.value = value
+        self.is_leaf = is_leaf
+        if is_leaf:
+            self.children = None
+        else:
+            self.children = {}
+
+    def __contains__(self, item):
+        return item in self.children
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py]: return (item in self.children) if self.children else False?

File: sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+    def get(self, key):
+        return self.children.get(key)
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py]: same to the above nit when self.children is None

File: sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+    def put(self, key, value):
+        self.children[key] = value
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py]: same to the above nit when self.children is None

File: sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    def __init__(
+        self
+    ):
+        self.ownerships_trie = _DictTrie(
+            "ownerships_trie",
+            path_names=("fully_qualified_namespace", "eventhub_name", "consumer_group", "partition_id")
+        )
+        self.checkpoints_trie = _DictTrie(
+            "checkpoints_trie",
+            path_names=("fully_qualified_namespace", "eventhub_name", "consumer_group", "partition_id")
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py]: should ownerships_trie and checkpoints_trie  be private?
```
self._ownership_trie
self._checkpoints_trie
```

File: sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py (hunk 1 above)
Comment (YijunXieMS) [reply sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py]: Changed to raise an error if calling it with a leaf node

File: sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py (hunk 3 above)
Comment (YijunXieMS) [reply sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py]: Changed to raise an error if calling it with a leaf node

File: sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py (hunk 4 above)
Comment (YijunXieMS) [reply sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py]: Changed to it

File: sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+            raise ValueError("Please don't put a value to a leaf TrieNode")
+        self.children[key] = value
+
+
+def _lookup_trie(root, path, path_index):
+    if path_index == len(path) - 1:
+        return root.get(path[path_index])
+    if path[path_index] in root:
+        return _lookup_trie(root.children[path[path_index]], path, path_index+1)
+    return None
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py]: ```key_name = path[path_index]``` for readability

File: sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+def _set_ele_trie(root, ele, path_names, path_index):
+    if path_index == len(path_names) - 1:
+        root.children[ele[path_names[path_index]]] = _TrieNode(ele, True)
+        return
+    next_node = root.get(ele[path_names[path_index]])
+    if not next_node:
+        next_node = _TrieNode(ele[path_names[path_index]], False)
+        root.put(ele[path_names[path_index]], next_node)
+    _set_ele_trie(next_node, ele, path_names, path_index + 1)
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/in_memory_checkpoint_store.py]: ```key_name = path_names[path_index]```

Nit: `def _lookup_trie(root, path, path_index):` has param `path`, while here it is named `path_names`

File: sdk/eventhub/azure-eventhub/azure/eventhub/aio/_eventprocessor/in_memory_checkpoint_store.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# -----------------------------------------------------------------------------------
+from typing import List, Dict, Any, Iterable, Optional, Union
+from azure.eventhub._eventprocessor.in_memory_checkpoint_store import InMemoryCheckpointStore as CheckPointStoreImpl
+from .checkpoint_store import CheckpointStore
+
+
+class InMemoryCheckpointStore(CheckpointStore):
+    def __init__(self):
+        self.checkpoint_store_impl = CheckPointStoreImpl()
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/aio/_eventprocessor/in_memory_checkpoint_store.py]: private variable: ```self._checkpoint_store_impl```

Comment (YijunXieMS) [reply sdk/eventhub/azure-eventhub/azure/eventhub/aio/_eventprocessor/in_memory_checkpoint_store.py]: changed

**Pull request merged**